jax/core.py

```diff
@@ -736,6 +736,9 @@ def reset_trace_state() -> bool:
 def cur_sublevel() -> Sublevel:
   return thread_local_state.trace_state.substack[-1]
 
+def cur_level() -> int:
+  return thread_local_state.trace_state.trace_stack.stack[-1].level
```

Comment:
As discussed on chat, users will need a little more than this to detect all the things. In deepmind/dm-haiku#301 I've checked in a version that reaches into core to do this (we will update as soon as this PR is merged and a stable release of JAX is made including it). Our impl:

```python
Opaque = Any

def jax_trace_level(cls) -> Opaque:
  trace_stack = jax_core.thread_local_state.trace_state.trace_stack.stack
  top_type = trace_stack[0].trace_type
  level = trace_stack[-1].level
  sublevel = jax_core.cur_sublevel()
  return top_type, level, sublevel
```